proc/compiler/groupby.go

```diff
@@ -0,0 +1,103 @@
+package compiler
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/brimsec/zq/ast"
+	"github.com/brimsec/zq/expr"
+	"github.com/brimsec/zq/field"
+	"github.com/brimsec/zq/proc"
+	"github.com/brimsec/zq/proc/groupby"
+	"github.com/brimsec/zq/reducer"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+func compileGroupBy(pctx *proc.Context, parent proc.Interface, node *ast.GroupByProc) (*groupby.Proc, error) {
+	keys, err := compileKeys(node.Keys, pctx.TypeContext)
+	if err != nil {
+		return nil, err
+	}
+	names, reducers, err := compileReducers(node.Reducers, pctx.TypeContext)
+	if err != nil {
+		return nil, err
+	}
+	return groupby.New(pctx, parent, keys, names, reducers, node.Limit, node.InputSortDir, node.ConsumePart, node.EmitPart)
+}
+
+func compileKeys(assignments []ast.Assignment, zctx *resolver.Context) ([]expr.Assignment, error) {
+	keys := make([]expr.Assignment, 0, len(assignments))
+	for _, assignment := range assignments {
+		a, err := expr.CompileAssignment(zctx, &assignment)
+		if err != nil {
+			return nil, err
+		}
+		keys = append(keys, a)
+	}
+	return keys, nil
+}
+
+func compileReducers(assignments []ast.Assignment, zctx *resolver.Context) ([]field.Static, []reducer.Maker, error) {
+	names := make([]field.Static, 0, len(assignments))
+	reducers := make([]reducer.Maker, 0, len(assignments))
+	for _, assignment := range assignments {
+		name, f, err := compileReducer(zctx, assignment)
+		if err != nil {
+			return nil, nil, err
+		}
+		reducers = append(reducers, f)
```

Comment:
NIt: f is surprising name for a reducer.Maker.
```suggestion
		name, maker, err := compileReducer(zctx, assignment)
		if err != nil {
			return nil, nil, err
		}
		reducers = append(reducers, maker)
```